Attach Cloudflare Zero Trust headers in API calls to Baby Buddy server

btoconnor opened the pull request:
Issue #7 

Allow setting of (optional) [Cloudflare Zero Trust](https://developers.cloudflare.com/cloudflare-one/) headers so that the Echo devices can talk to a Baby Buddy instance protected by CF Zero Trust using a [Service Token](https://developers.cloudflare.com/cloudflare-one/identity/service-tokens/).

*Description of changes:*
Attach `CF-Access-Client-Id` and `CF-Access-Client-Secret` headers to API requests from the Alexa skill to the baby buddy server.

This allows the user to specify the headers in the `bin/setup.sh` script (though in the future, presumably this should be optional, probably through some command line argument if it were to be merged).


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

cdubz commented on bin/setup.sh:
```diff
@@ -29,10 +29,20 @@ if [[ ! -f $CONFIG_FILE ]]; then
 
   read BABY_BUDDY_API_KEY
 
+  echo -n "Enter CF Client Id: "
+
+  read CF_ACCESS_CLIENT_ID
+
+  echo -n "Enter CF Client Secret: "
+
+  read CF_ACCESS_CLIENT_SECRET
+
```
Can we make the Cloudflare questions branching? E.g., ask first "Is the Baby Buddy instance protected by CloudFlare Zero Trust?". If yes, branch to the CF header questions and if no skip them (and don't add them to the config).

cdubz commented on lambda/custom/src/babybuddy/api.ts:
```diff
@@ -13,10 +13,12 @@ import {
 const fetchSecrets: () => Promise<Secret> = async () => {
   return new Promise((resolve, reject) => {
-    if (process.env.BABY_BUDDY_API_KEY && process.env.BABY_BUDDY_API_URL) {
+    if (process.env.BABY_BUDDY_API_KEY && process.env.BABY_BUDDY_API_URL && process.env.CF_ACCESS_CLIENT_ID && process.env.CF_ACCESS_CLIENT_SECRET) {
       resolve({
         apiKey: process.env.BABY_BUDDY_API_KEY,
         apiUrl: process.env.BABY_BUDDY_API_URL,
+        cfAccessClientId: process.env.CF_ACCESS_CLIENT_ID,
+        cfAccessClientSecret: process.env.CF_ACCESS_CLIENT_SECRET
```
Assuming the CF header variables can be made optional -- lets keep the current structure but but the object to go to `resolve`, add the CF headers if provided, and then call `resolve` at the end.

cdubz commented on lambda/custom/src/babybuddy/api.ts:
```diff
@@ -26,14 +28,16 @@ const fetchSecrets: () => Promise<Secret> = async () => {
+    const { apiKey, apiUrl, cfAccessClientId, cfAccessClientSecret } = await fetchSecrets();
 
     try {
       const response = await axios.get(url, {
         baseURL: apiUrl,
         headers: {
           'Content-Type': 'application/json',
           Authorization: `Token ${apiKey}`,
+          'CF-Access-Client-Id': `${cfAccessClientId}`,
+          'CF-Access-Client-Secret': `${cfAccessClientSecret}`
         },
       });
```
Lets create a separate base function for requests. In all three cases (`get`, `post`, and `delete`) the options passed to axios are the same:

```js
{
        baseURL: apiUrl,
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Token ${apiKey}`,
          'CF-Access-Client-Id': `${cfAccessClientId}`,
          'CF-Access-Client-Secret': `${cfAccessClientSecret}`
        },
      }
```

So those options can established in the base function and the CF headers can be added conditionally only if set.

cdubz commented on lambda/custom/src/babybuddy/types.ts:
```diff
@@ -66,6 +66,8 @@ interface CreateDiaperChange {
 interface Secret {
   apiKey: string;
   apiUrl: string;
+  cfAccessClientId: string;
+  cfAccessClientSecret: string;
```
More TS ignorance here... assuming making this optional is possible I suspect something about this interface declaration would need to be updated?

cdubz reviewed (changes requested):
Bear with me here as I'm not well versed in either TS or Alexa... my review basically suggests that we make the CloudFlare headers _optional_ to support this.

btoconnor commented:
> Bear with me here as I'm not well versed in either TS or Alexa... my review basically suggests that we make the CloudFlare headers _optional_ to support this.

Definitely can make it optional.  My initial PR was just my hacked up code so that I could get it to work.  I just created a commit that I believe makes the Cloudflare header support optional.  I _believe_ I addressed the points you made in your initial review but let me know if you've got questions or any other concerns!

Thanks for the quick responses / review.

cdubz reviewed (approved):
This looks good to me. @rplankenhorn are you available to review as well?

btoconnor commented:
@cdubz any thoughts on this?

cdubz commented:
I was hoping @rplankenhorn could pop in and review but I will keep this in my queue and follow up soonish.

cdubz commented:
Ok after much fiddling I was able to get CZFT working with a test instance of Baby Buddy 🎉 

This all looks good but it definitely _is_ a hyper-specific use case. I think an ideal future state would be a configuration process allowing the user to define any number of arbitrary headers to send with API request.

An end result `.env` file could look like:

```env
BABY_BUDDY_API_KEY=KEY
BABY_BUDDY_API_URL=URL
EXTRA_HEADERS=CF-Access-Client-Id,CF-Access-Client-Secret
EXTRA_HEADERS_VALUES=CLIENT_ID,SECRET
```

Maybe a comma-separated list of items would be limiting in some way but something like that.

Anyway! This Alexa skill itself is enough of a niche that I have no problem merging this as-is.

btoconnor commented:
Definitely agree. I wouldn't want to add service specific configuration for every conceivable option in the future.

If this comes up again, I was thinking something like a json structure of header / values would be preferred. Can tackle that in the future when it comes up.

Thanks for taking a look! Will hopefully get the next patches lined up soon
